src/middleware/legacy-routes-redirect.ts

```diff
@@ -0,0 +1,27 @@
+import { redirect } from "@solidjs/router";
+import { type FetchEvent } from "@solidjs/start/server/types";
+
+/**
+ * Redirect Map
+ * [origin, destination]
+ */
+const LEGACY_ROUTES = new Map([
+	[
+		"/references/api-reference/control-flow/Suspense",
+		"/reference/components/suspense",
+	],
+] as const);
+
+function isLegacyRoute(
+	path: string
+): path is Parameters<(typeof LEGACY_ROUTES)["get"]>[number] {
+	return LEGACY_ROUTES.has(path);
+}
+
+export const handleLegacyRoutes = (event: FetchEvent) => {
+	const { pathname } = new URL(event.request.url);
+
+	if (isLegacyRoute(pathname)) {
+		return redirect(LEGACY_ROUTES.get(pathname)!, 308);
+	}
+};
```

Comment:
If you use an object instead of a `Map`, you can avoid a lot of types fighting:

```tsx
import { redirect } from "@solidjs/router";
import { type FetchEvent } from "@solidjs/start/server/types";

/**
 * Redirect Map
 * {origin: destination}
 */
const LEGACY_ROUTES = {
	"/references/api-reference/control-flow/Suspense": "/reference/components/suspense"
}

export const handleLegacyRoutes = (event: FetchEvent) => {
	const { pathname } = new URL(event.request.url);

	if (pathname in LEGACY_ROUTES) {
		return redirect(LEGACY_ROUTES[pathname], 308);
	}
};
```

Reply:
used your solution. But couldn't get rid of the type predicate, since `pathname in LEGACY_ROUTES` does not narrows `pathname` as a key of  `LEGACY_ROUTES`